File: src/modules/chat/schemas.ts
import { z } from "zod";
import { ToolAction } from "@/modules/agent/enums";

export enum ChatRequestType {
  Message = "message",
  ToolAction = "tool_action",
}

const SendMessageSchema = z.object({
  type: z.literal(ChatRequestType.Message),
  threadId: z.string().optional(),
  content: z.string().min(1),
  context: z.record(z.string(), z.unknown()).optional(),
});

const ToolActionSchema = z.object({
  type: z.literal(ChatRequestType.ToolAction),
  threadId: z.string(),
  toolCallId: z.string(),
  action: z.enum(ToolAction),
  modifiedArgs: z.record(z.string(), z.unknown()).optional(),
});

export const ChatRequestSchema = z.discriminatedUnion("type", [
  SendMessageSchema,
  ToolActionSchema,
]);

export type ChatRequest = z.infer<typeof ChatRequestSchema>;
export type SendMessageRequest = z.infer<typeof SendMessageSchema>;
export type ToolActionRequest = z.infer<typeof ToolActionSchema>;

File: app/modules/chat/schemas.py
"""Pydantic schemas for chat API (Vercel AI SDK–aligned)."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional, Union

from pydantic import BaseModel, ConfigDict, Field

from app.common.enums import MessagePartType, MessageRole


# --- Message parts (discriminated union by type) ---


class TextPart(BaseModel):
    type: Literal[MessagePartType.TEXT] = MessagePartType.TEXT
    text: str = ""


class ToolInvocationPart(BaseModel):
    type: Literal[MessagePartType.TOOL_INVOCATION] = MessagePartType.TOOL_INVOCATION
    tool_call_id: str = Field("", alias="toolCallId")
    tool_name: str = Field("", alias="toolName")
    state: Optional[str] = None
    args: Optional[Dict[str, Any]] = None
    result: Any = None

    model_config = ConfigDict(populate_by_name=True)


class ToolResultPart(BaseModel):
    type: Literal[MessagePartType.TOOL_RESULT] = MessagePartType.TOOL_RESULT
    tool_call_id: str = Field("", alias="toolCallId")
    tool_name: str = Field("", alias="toolName")
    result: Any = None
    action: Optional[str] = None  # approved | rejected | cancelled
    is_approval: bool = Field(False, alias="isApproval")

    model_config = ConfigDict(populate_by_name=True)


MessagePart = Union[TextPart, ToolInvocationPart, ToolResultPart]


class UIMessage(BaseModel):
    """Single UI message (Vercel AI SDK useChat)."""

    id: str = ""
    role: MessageRole
    content: str = ""
    parts: List[Union[TextPart, ToolInvocationPart, ToolResultPart]] = Field(
        default_factory=list
    )


class ChatRequest(BaseModel):
    """POST /chat body: messages and optional thread_id."""

    thread_id: Optional[str] = Field(None, alias="threadId")
    messages: List[UIMessage]
    context: Optional[Dict[str, Any]] = None

    model_config = ConfigDict(populate_by_name=True)
